runtime/src/iree/hal/utils/libmpi.c

```diff
@@ -67,21 +69,39 @@ iree_status_t iree_hal_mpi_library_load(
 // convert an MPI_Status result to an iree_status_t
 //
 // We convert MPI_SUCCESS to IREE_STATUS_SUCCESS, and MPI errors to
-// IREE_STATUS_INTERNAL and print the MPI error code.
-//
-// TODO(cascaval): use MPI_Error to print the actual error or consider
-// mapping to other error codes.
+// IREE_STATUS_INTERNAL and retrieve the error string from the MPI library.
 iree_status_t iree_hal_mpi_result_to_status(
     iree_hal_mpi_dynamic_symbols_t* syms, int result, const char* file,
     uint32_t line) {
-  iree_status_code_t code;
-  switch (result) {
-    case 0:  // MPI_SUCCESS
-      return iree_ok_status();
-    default:
-      code = IREE_STATUS_INTERNAL;
-      break;
+  // the MPI spec mandates MPI_SUCCESS == 0, all the other errors are
+  // implementation dependent. However, since we already depend on OpenMPI
+  // for implementation dependent handles, we can hardcode the error classes
+  // as well.
+  const int MPI_SUCCESS = 0;
+  const int MPI_ERR_UNKNOWN = 14;
```

Comment:
We should be mapping the error classes back to our own status codes instead of always mapping to INTERNAL, if they make sense.
![image](https://user-images.githubusercontent.com/75337/235519514-9929374b-03e1-483d-9fac-4561237bb09c.png)
Here most of these could map to IREE_STATUS_INVALID_ARGUMENT, TRUNCATE -> DATA_LOSS, INTERN -> INTERNAL, UNKNOWN -> UNKNOWN, PENDING -> FAILED_PRECONDITION (precondition being that the request was not pending)

Reply:
Agreed! Unfortunately the standard only defines MPI_SUCCESS == 0, all the other errors are implementation dependent. For example:

| Error class         |  OpenMPI   | MPICH  |
-----------------------|-----------------|------------|
MPI_SUCCESS                  |  0         | 0 
MPI_ERR_BUFFER               |  1         | 1
MPI_ERR_COUNT                |  2         |  2
MPI_ERR_TYPE                 |  3         |  3 
MPI_ERR_TAG                  |  4         |  4
MPI_ERR_COMM                 |  5         | 5
MPI_ERR_RANK                 |  6         |  6 
MPI_ERR_REQUEST              |  7         | 19
MPI_ERR_ROOT                 |  8         | 7
MPI_ERR_GROUP                |  9         | 8
MPI_ERR_OP                   |  10         | 9
MPI_ERR_TOPOLOGY             |  11         | 10
MPI_ERR_DIMS                 |  12         | 11 
MPI_ERR_ARG                  |  13         | 12 
MPI_ERR_UNKNOWN              |  14         | 13
MPI_ERR_TRUNCATE             |  15         | 14
MPI_ERR_OTHER                |  16         | 15
MPI_ERR_INTERN               |  17         | 16
MPI_ERR_IN_STATUS            |  18         | 17
MPI_ERR_PENDING              |  19         | 18
MPI_ERR_ACCESS               |  20         | 20
MPI_ERR_AMODE                |  21         | 21
MPI_ERR_ASSERT               |  22         | 53
MPI_ERR_BAD_FILE             |  23         | 22
MPI_ERR_BASE                 |  24         | 46
MPI_ERR_CONVERSION           |  25         | 23
MPI_ERR_DISP                 |  26         |  52
MPI_ERR_DUP_DATAREP          |  27         | 24
MPI_ERR_FILE_EXISTS          |  28         | 25
MPI_ERR_FILE_IN_USE          |  29         | 26
MPI_ERR_FILE                 |  30         | 27
MPI_ERR_INFO_KEY             |  31         | 29
MPI_ERR_INFO_NOKEY           |  32         | 31
MPI_ERR_INFO_VALUE           |  33         | 30
MPI_ERR_INFO                 |  34         | 28
MPI_ERR_IO                   |  35         | 32
MPI_ERR_KEYVAL               |  36         | 48
MPI_ERR_LOCKTYPE             |  37         | 47
MPI_ERR_NAME                 |  38         | 33
MPI_ERR_NO_MEM               |  39         | 34
MPI_ERR_NOT_SAME             |  40         | 35 
MPI_ERR_NO_SPACE             |  41         | 36
MPI_ERR_NO_SUCH_FILE         |  42         |37
MPI_ERR_PORT                 |  43         | 38
MPI_ERR_QUOTA                |  44         | 39
MPI_ERR_READ_ONLY            |  45         |40
MPI_ERR_RMA_CONFLICT         |  46         |49
MPI_ERR_RMA_SYNC             |  47         |50
MPI_ERR_SERVICE              |  48         |41
MPI_ERR_SIZE                 |  49         |51
MPI_ERR_SPAWN                |  50         |42
MPI_ERR_UNSUPPORTED_DATAREP  |  51         | 43
MPI_ERR_UNSUPPORTED_OPERATION|  52         |44
MPI_ERR_WIN                  |  53         | 45
... | |
MPI_ERR_LASTCODE             |  92 | 79

So unless we figure out a way to compile using mpi.h with symbolic values for these errors, we'll have to replicate the definitions for each implementation. Which in my view, it is not maintainable.

Or leave them as is.

Reply:
That's the worst. Why even have a spec... 🤦